tools/istio-iptables/pkg/dependencies/implementation.go

```diff
@@ -130,47 +257,28 @@ const (
 // 4. Otherwise, see if we have `legacy` binary set, and use that.
 // 5. Otherwise, see if we have `iptables` binary set, and use that (detecting whether it's nft or legacy).
 func (r *RealDependencies) DetectIptablesVersion(ipV6 bool) (IptablesVersion, error) {
-	// Begin detecting
-	//
-	// iptables variants all have ipv6 variants, so decide which set we're looking for
-	var nftBin, legacyBin, plainBin string
-	if ipV6 {
-		nftBin = ip6tablesNftBin
-		legacyBin = ip6tablesLegacyBin
-		plainBin = ip6tablesBin
+	modules, err := LoadProcModules()
+	if err != nil {
+		log.Errorf("failed to load kernel modules from %s, error was: %s", ModulesFile, err.Error())
+		return IptablesVersion{}, err
+	}
+	legacyVersion, err := GetLegacyVersion(modules, ipV6)
+	nftVersion, err := GetNFTVersion(modules, ipV6)
+	if legacyVersion != nil && nftVersion != nil {
+		if legacyVersion.ExistingRules {
+			log.Info("legacy and nft both exist but nft has rules, so use legacy")
+			return *legacyVersion, nil
+		}
+		log.Info("legacy and nft both exist but legacy does not have any rules, use nft by default")
+		return *nftVersion, nil
 	} else {
-		nftBin = iptablesNftBin
-		legacyBin = iptablesLegacyBin
-		plainBin = iptablesBin
-	}
-
-	// 1. What binaries we have
-	// 2. What binary we should use, based on existing rules defined in our current context.
-	// does the nft binary set exist, and are nft rules present?
-	nftVer, err := shouldUseBinaryForCurrentContext(nftBin)
-	if err == nil && nftVer.ExistingRules {
-		// if so, immediately use it.
-		return nftVer, nil
-	}
-	// not critical, may find another.
-	log.Debugf("did not find (or cannot use) iptables binary, error was %w: %+v", err, nftVer)
-
-	// Check again
-	// does the legacy binary set exist, and are legacy rules present?
-	legVer, err := shouldUseBinaryForCurrentContext(legacyBin)
-	if err == nil && legVer.ExistingRules {
-		// if so, immediately use it
-		return legVer, nil
-	}
-	// not critical, may find another.
-	log.Debugf("did not find (or cannot use) iptables binary, error was %w: %+v", err, legVer)
-
-	// regular non-suffixed binary set is our last resort.
-	//
-	// If it's there, and rules do not already exist for a specific variant,
-	// we should use the default non-suffixed binary.
-	// If it's NOT there, just propagate the error, we can't do anything, no iptables here
-	return shouldUseBinaryForCurrentContext(plainBin)
+		if nftVersion != nil {
+			log.Info("use nft")
+			return *nftVersion, nil
+		}
+		log.Info("use legacy")
+		return *legacyVersion, nil
+	}
```

Comment:
AFAICT all that needs to change from the existing code to fix #52357 is to prefer `nft` if both binaries are available and rules are defined in `nft` AND `legacy` (or in neither), rather than early-exiting on line 153?